std/array.d

```diff
@@ -332,6 +335,97 @@ ElementType!String[] array(String)(String str) if (isNarrowString!String)
     )));
 }
 
+
+/**
+ * Return a static array filled with elements from a range.
+ *
+ * Params:
+ *      range = a range (or aggregate with $(D opApply) function) whose elements are copied into the static array
+ * Returns:
+ *      initialized static array
+*/
+auto staticArray(alias range)()
+if (isIterable!(typeof(range)) && !isInfinite!(typeof(range)))
+{
+    import std.array : array;
+    enum arr = array(range);
+    enum sA = .staticArray(arr);
+    return sA;
+}
+
+///
+@safe @nogc nothrow unittest
+{
+    import std.range;
+    enum r = iota(3);
+    auto sa4 = staticArray!r;
+    static assert(sa4.length == 3);
+    assert(sa4 == [0, 1, 2]);
+}
+
+/**
+ * Returns a static array filled with elements copied from a dynamic array.
+ * In contrast to the overload that takes its source as a template alias, this takes
+ * a it as a normal argument. This allows it to be used with runtime arrays in
+ * situations where the length of the array is guaranteed in the function call expression
+ * (see examples for details).
+ *
+ * Optionally, a target type can be provided as the first template argument, which will result
+ * in a static array with that element type.
+*/
+T[N] staticArray(size_t N, T)(T[N] a)
+{
+    return a;
+}
+
+/// ditto
+TargetElemT[N] staticArray(TargetElemT, size_t N, T)(T[N] a)
+{
+    Unqual!(TargetElemT)[N] result = void;
+
+    import std.conv : emplaceRef;
+    foreach (i; 0 .. N)
+        emplaceRef!TargetElemT(result[i], a[i]);
```

Comment:
If any of these throws, the destructors of the already-constructed elements is not called.